Create AttributeSet for use by Nodes and Ports

TysonAndre opened the pull request:
Not doing verification against list of possible attribute names and
types yet.

Fixes  #81

TysonAndre committed:
Create AttributeSet for use by Nodes and Ports #81

Not doing verification against list of possible attribute names and
types yet.

mtrberzi commented:
I think this is a good start. This will be a useful class to have when attribute type verification does become a thing, which now becomes trivial because this class could just take a `Map<String, Type>` in order to know what attributes are allowed and what type each value should have.

lucaswoj commented:
Right now, this is a simple wrapper around a `Map` -- it'll be much more interesting when combined with #27 and #80. I'm ok merging this, pending a few quick changes requested in the comments. Lets start on those two tasks next!

lucaswoj commented on src/main/java/org/whdl/intermediate/AttributeSet.java:
```diff
@@ -0,0 +1,25 @@
+    data = new HashMap<String, Value>();
+  }
+
+  public Value get(String attrName) throws UndeclaredIdentifierException {
```
We might want a new exception type for this case. Thoughts?

lucaswoj commented on src/main/java/org/whdl/intermediate/AttributeSet.java:
```diff
@@ -0,0 +1,25 @@
+    if (data.containsKey(attrName)) {
+      return data.get(attrName);
+    } else {
+      throw new UndeclaredIdentifierException("no attribute named '" + attrName
```
We should not error pass strings directly into exceptions like this. Not your fault, Tyson, I'll make new issues to address the problem.

lucaswoj replied:
Instantiating `UndeclaredIdentifierException` takes the signature `public UndeclaredIdentifierException(String identifier)` so, in theory, you should just be passing `attrName`. I still think we should create a new exception type for this case but I don't have  a strong opinion.

lucaswoj commented on src/main/java/org/whdl/intermediate/AttributeSet.java:
```diff
@@ -0,0 +1,25 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class AttributeSet {
```
How about just calling this class `Attributes`? Its not _exactly_ a set.

mtrberzi commented:
Can this also be implemented on `Connection`s and `Constraint`s? It would be better for the sake of completeness to deal with all of these things at the same time.

TysonAndre committed:
Rename to Attributes, use in more places

Updating Connection and Constraint to use Attributes abstraction.

updating unit tests as well.

TysonAndre commented:
Made the suggested fixes.
- renamed AttributeSet to Attributes
- added UndeclaredAttributeException
- changed Constraint/Connection to use Attributes object

Probably should have merged against master while reviewing, then rebased before committing instead of rebasing early on.

mtrberzi commented:
Looks good to me. Merge!
